### DISCUSSION TITLE
Several differences in Book and Codes

### DISCUSSION THREAD

**Discussion by ShervanN** [ProfessionalCSharp/ProfessionalCSharp2021#177] [General]
Hi,

I am not sure but I've seen some differences(or mistakes?)

1) On page 461, The code is written:

```
private static void ThreadingTimer()
{
void TimeAction(object? o) =>
Console.WriteLine($"System.Threading.Timer {DateTime.Now:T}");
using Timer t1 = new(
TimeAction,
null,
dueTime: TimeSpan.FromSeconds(2),
period: TimeSpan.FromSeconds(3)))
{
Task.Delay(15000).Wait();
}
}
```

Does the above code have a compiler error? 

2) The code in [THIS](https://github.com/ProfessionalCSharp/ProfessionalCSharp2021/blob/main/2_Libs/Parallel/ParallelSamples/TimersSample/Program.cs) link  has:
`await Task.Delay(15000);`

But in the book, you wrote:
`Task.Delay(15000).Wait();
`

**Comment by christiannagel** [comment]
With the book printed code, there are too many curly braces. This should be with the **using declaration**.

```csharp
private static void ThreadingTimer()
{
  void TimeAction(object? o) =>
    Console.WriteLine($"System.Threading.Timer {DateTime.Now:T}");

  using Timer t1 = new(
    TimeAction,
    null,
    dueTime: TimeSpan.FromSeconds(2),
    period: TimeSpan.FromSeconds(3));

  Task.Delay(15000).Wait();
}
```

I've updated [updates.md](https://github.com/ProfessionalCSharp/ProfessionalCSharp2021/blob/main/Updates.md)

As an alternative way - with the **using statement** this could be done:

```csharp
private static void ThreadingTimer()
{
  void TimeAction(object? o) =>
    Console.WriteLine($"System.Threading.Timer {DateTime.Now:T}");

  using (Timer t1 = new(
    TimeAction,
    null,
    dueTime: TimeSpan.FromSeconds(2),
    period: TimeSpan.FromSeconds(3)))
  {
    Task.Delay(15000).Wait();
  }
}
```

Before C# 8 only the *using statement* was possible. See [Using, using, using with C# 8](https://csharp.christiannagel.com/2019/04/09/using/) on differences.

With a previous C# version, it was also not possible to use `await` within the `Main` method. This is possible now, and the `await` can also be used with top-level statements. The current version from the book got rid of the method and just uses top-level statements:

```csharp
static void TimeAction(object? o) =>
    Console.WriteLine($"System.Threading.Timer {DateTime.Now:T}");

using Timer t1 = new(
    TimeAction, 
    null, 
    dueTime: TimeSpan.FromSeconds(2),
    period: TimeSpan.FromSeconds(3));

await Task.Delay(15000);
```

Remember, the `Wait` method blocks the calling thread, while with the `await` the thread is not blocked, but can do something else. With this simple console application it doesn't matter which one to use.

**Reply by ShervanN** [reply]
Much obliged.
